### PR TITLE
nixos/stunnel: Make free-form

### PR DESCRIPTION
**Pull request by chkno**
###### Motivation for this change
Allow the use of stunnel functionality beyond the very small (5%) subset explicitly supported by the NixOS configuration schema.
Also, add some tests of core stunnel functionality.

This supersedes PR #96401, which proposed just adding a few more config directives rather than a general escape hatch.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
    - I did this, but I noticed that `nix-build . -A stunnel.tests` _doesn't_ run the tests as the linked documentation suggests it should.  `nix-build . -A stunnel.tests.stunnel` does run the tests.
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

FYI: stunnel maintainer @thoughtpolice, stunnel NixOS module author @lschuermann.

### CODE REVIEW

File: nixos/tests/stunnel.nix
Comment (martinetd) [inline nixos/tests/stunnel.nix:100]: (nitpick) might make sense to use the same copy code for first test and have `curl --cafile /authorized-server-cert.crt` instead of `--insecure` ? but to be honest the second tests's client stunnel will verify the server certificate so it doesn't really matter imo.

**Review by martinetd — APPROVED** [review]
extraConfig never hurts... And adding tests is great, thanks!

File: nixos/tests/stunnel.nix
Comment (chkno) [reply nixos/tests/stunnel.nix]: Good idea.  Done.  Thanks!

**Review by martinetd — APPROVED** [review]
cheers!
I ran the new tests and confirmed it all works.

**Review by lschuermann — APPROVED** [review]
LGTM.

**Comment by mohe2015** [comment]
It's not possible to do this via a freeform module? Because I think that's preferred

File: nixos/tests/stunnel.nix
Comment (astro) [inline nixos/tests/stunnel.nix:75]: maybe make it more obvious that this is the client, eg.
```
# client
```
or rename it

File: nixos/tests/stunnel.nix
Comment (astro) [inline nixos/tests/stunnel.nix:92]: `server`

**Review by astro — APPROVED** [review]

**Comment by chkno** [comment]
@mohe2015 - freeform module - Ah!  Yes!  That is definitely worth trying.  Thanks!  I'm flipping this PR to draft now & will be right back with that.

**Comment by chkno** [comment]
Per @mohe2015 's suggestion, this PR now converts stunnel into a free-form module, rather than merely adding text-based extraConfig escape hatches.

This is now a more invasive change.  I believe I've preserved backward compatibility: I added tests for the existing functionality in a separate commit before changing anything.

Please have another look.  Thanks!

**Review by martinetd — COMMENTED** [review]
hi! I've just been looking at converting something else to freeform and was pointed out at this: https://github.com/NixOS/rfcs/pull/42

long story short, it looks like there's a generic way of creating freeform modules now instead of your generateConfig -- I'm not sure if that's important though as long as it works? *clueless*

that aside it looks good at first glance, I'd need to actually try a few old configs to convince myself it's mostly compatible but need to find some more time to do that..

EDIT: in particular for freeform things you could probably use pkgs.formats.ini or toml as is? it looks like a good match

**Comment by chkno** [comment]
@martinetd - Unfortunately, the stunnel configuration format is [kind of a mess](https://github.com/mtrojnar/stunnel/blob/master/src/options.c).  It is much less sophisticated than INI or TOML.  For example, it doesn't have any notion of quoting and escaping.  Given a section name with a ] in it, the INI generator will escape it: `[The \\] Section]'.  Stunnel doesn't parse this way, and generating configuration as if it would be parsed this way would be confusing to users.  The INI generator can be configured not to do this, and configured to use "yes" and "no" instead of "true" and "false", and  we could remove keys with null values before handing off to the generator, and so on, but by that point it's just as much configuration-code to interface with the generator library as to just generate it directly, and requires future maintainers to read much more configuration-code to understand what's going on.

I added a brief note to the configuration generator with this rationale.

**Comment by chkno** [comment]
@martinetd, @chkno -- Eh, actually it's not so bad.  Now updated to use `generators.toINI`.  Thanks!

**Review by martinetd — APPROVED** [review]
Sorry, didn't mean to make you do so much work - I was happy by a config is messed up the generator code wasn't that bad... But well ok thanks ;)

I've read through it (looks fine) and verified it doesn't break my simple setup. Thanks!

**Comment by chkno** [comment]
Thanks, [martinetd](https://github.com/martinetd)!

Any other reviewers have any feedback?

**Comment by chkno** [comment]
I'd like to progress this toward merge.  The [PRs already reviewed thread](https://discourse.nixos.org/t/prs-already-reviewed/2617) says that it is for _reviewers_ to submit things there, not PR _authors_, so can reviewers either link this PR there or give feedback here about what is needed before this PR can be merged?  Thank you!

**Pull request merged**
